File: cpp/src/statechart/state.cpp
/*
 * UML statechart framework (http://uml-statecharts.sourceforge.net)
 *
 * Copyright (C) 2006-2007 Christian Mocek (klangfarbe@users.sourceforge.net)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
 */
#include <statechart/state.h>
#include <statechart/context.h>
#include <statechart/statechart.h>
#include <statechart/timeoutevent.h>
#include <statechart/statechartexception.h>
#include <stdlib.h>
#include <string.h>

using namespace std;
using namespace statechart;
using namespace statechart::util;

//------------------------------------------------------------------------------

State::State(Context* parent, Action* entryAction, Action* doAction, 
             Action* exitAction, const char* name) {
  this->context = parent;
  this->statechart = 0;

  // get the statechart, we need it for the event queues
  if(parent) {
    parent->addSubstate(this);
    while(parent->context) {
      parent = parent->context;
    }
    statechart = dynamic_cast<Statechart*>(parent);
    if(!statechart) {
      throw new StatechartException("Cannot determine path to the statechart. Check the hierarchy.");
    }
  }

  this->entryAction = entryAction;
  this->doAction = doAction;
  this->exitAction = exitAction;
  this->name = 0;
  setName(name);
}

//---------------------------------------------------------------------------

State::~State() {
  if(name) {
    free(name);
  }

  delete entryAction;
  delete exitAction;
  delete doAction;

  // delete Transisions
  for(unsigned int i = 0; i < transitions.size(); i++) {
    delete transitions[i];
    transitions[i] = 0;
  }
  transitions.clear();
}

//---------------------------------------------------------------------------

void State::setEntryAction(Action* action) {
  delete entryAction;
  entryAction = action;
}

//---------------------------------------------------------------------------

void State::setDoAction(Action* action) {
  delete doAction;
  doAction = action;
}

//---------------------------------------------------------------------------

void State::setExitAction(Action* action) {
  delete exitAction;
  exitAction = action;
}

//---------------------------------------------------------------------------

bool State::activate(SmartPointer<Metadata>& data, 
                     SmartPointer<Parameter>& parameter) {
  if(!data->isActive(this)) {
    data->activate(this);

    // trigger the timout events if available
    unsigned long time = data->getData(this)->timestamp;
    for(unsigned int i = 0; i < transitions.size(); i++) {
      Transition* t = dynamic_cast<Transition*>(transitions[i]);
      if(t->hasEvent()) {
        TimeoutEvent* event = dynamic_cast<TimeoutEvent*>(t->event);
        if(event) {
          // Add an event with it's absolute time (stateactivation time + timeout)
          SmartPointer<Event> e(new TimeoutEvent(time + event->getTimeout()));
          statechart->timeoutEventQueue.add(this, data, e, parameter);
        }
      }
    }

    if(entryAction) {
      entryAction->execute(data, parameter);
    }

    if(doAction) {
      doAction->execute(data, parameter);
    }
    return true;
  }
  return false;
}

//---------------------------------------------------------------------------

void State::deactivate(SmartPointer<Metadata>& data, 
                       SmartPointer<Parameter>& parameter) {
  if(data->isActive(this)) {
    data->deactivate(this);
    if(exitAction) {
      exitAction->execute(data, parameter);
    }
  }
  statechart->timeoutEventQueue.remove(this, data);
}

//---------------------------------------------------------------------------

bool State::dispatch(SmartPointer<Metadata>& data, 
                     SmartPointer<Event>& event, 
                     SmartPointer<Parameter>& parameter) {
  for(unsigned int i = 0; i < transitions.size(); i++) {
    if(transitions[i]->execute(event, data, parameter)) {
      return true;
    }
  }
  return false;
}

//---------------------------------------------------------------------------

void State::addTransition(Transition* transition) {
  // make sure transition with guards are checked first!
  if(transition->hasGuard()) {
    transitions.insert(transitions.begin(), transition);
  } else {
    transitions.push_back(transition);
  }
}

//---------------------------------------------------------------------------

void State::setName(const char* _name) {
  if(this->name) {
    free(this->name);
    this->name = 0;
  }

  if(_name) {
    this->name = strdup(_name);
  }
}

File: src/statechart/State.java
/*
 * UML statechart framework (http://github.com/klangfarbe/UML-Statechart-Framework-for-Java)
 *
 * Copyright (C) 2006-2013 Christian Mocek (christian.mocek@gmail.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
 */
package statechart;

import java.util.Vector;

/**
 * Represents a simple state of the statechart. For debugging purposes each
 * state can have a name assigned for identification.
 */
public class State {
  //============================================================================
  // ATTRIBUTES
  //============================================================================
  // The entry action to execute.
  protected Action entryAction= null;

  // The do action to execute.
  protected Action doAction = null;

  // The exit action to execute.
  protected Action exitAction = null;

  // List of the associated transition objects.
  protected Vector<Transition> transitions = new Vector<Transition>();

  // The context of this state.
  Context context = null;
  
  // The statechart this state belongs to.
  Statechart statechart = null;

  // The name of the state (for easier debugging).
  protected String name = null;
  
  //============================================================================
  // METHODS
  //============================================================================
  /**
   * Creates a state with the given actions. 
   * @throws StatechartException 
   */
  public State(String name, Context parent, Action entryAction, Action doAction, Action exitAction) throws StatechartException {
    if(parent == null && !(this instanceof Statechart)) {
      throw new StatechartException("Parameter parent cannot be null");
    }
    if(name == null) {
      throw new StatechartException("Parameter name cannot be null");
    }
    
    this.context = parent;
    
    // get the statechart, we need it for the event queues
    if(parent != null) {
      while(parent.context != null) {
        parent = parent.context;
      }
      if(parent instanceof Statechart) {
        this.statechart = (Statechart)parent;
        if(this.statechart.states.containsKey(name)) {
          throw new StatechartException("State name <" + name + "> already used! Please define a unique name.");
        }
        this.statechart.states.put(name, this);
      } else {
        throw new StatechartException("Cannot determine path to the statechart. Check the hierarchy.");
      }
    }

    this.name = name;
    this.entryAction = entryAction;
    this.doAction = doAction;
    this.exitAction = exitAction;    
  }

  //============================================================================

  /**
   * Gets the context
   */
  public Context getContext() { 
    return context;
  };

  //============================================================================

  /**
   * Sets the entry action. If there is already an action given,
   * it will be destroyed first.
   */
  public void setEntryAction(Action action) {
    this.entryAction = action;
  }

  //============================================================================

  /**
   * Sets the do action. If there is already an action given,
   * it will be destroyed first.
   */
  public void setDoAction(Action action) {
    this.doAction = action;
  }

  //============================================================================

  /**
   * Sets the exit action. If there is already an action given,
   * it will be destroyed first.
   */
  public void setExitAction(Action action) {
    this.exitAction = action;    
  }

  //============================================================================

  /**
   * Sets the name of the state used for debugging purposes.
   */
  public void setName(String name) {
    this.name = name;
  }

  //============================================================================

  /**
   * return the given name
   */
  public String toString() { 
    return name == null ? "" : name;
  }

  //============================================================================

  /**
   * Adds a transition to the list.
   */
  void addTransition(Transition transition) {
    // make sure transition with guards are checked first!
    if(transition.hasGuard()) {
      transitions.add(0, transition);
    } else {
      transitions.add(transition);
    }
  }

  //============================================================================

  /**
   * Activates the state.
   */
  boolean activate(Metadata data, Parameter parameter) {
    if(!data.isActive(this)) {
      data.activate(this);
      
      // trigger the timout events if available
      for(int i = 0; i < transitions.size(); i++)  {
        Transition t = transitions.get(i);
        if(t.event != null && t.event instanceof TimeoutEvent) {
          TimeoutEvent event = (TimeoutEvent)t.event;          
          EventQueueEntry entry = new EventQueueEntry(statechart, 
                                                      this, data, event, parameter,
                                                      ((TimeoutEvent)event).getTimout());
          StateRuntimedata runtimedata = data.getData(this);
          runtimedata.timeoutEvents.add(entry);
          statechart.timeoutEventQueue.add(entry);
        }
      }
      
      if(entryAction != null) {
        entryAction.execute(data, parameter);
      }

      if(doAction != null) {
        doAction.execute(data, parameter);
      }
      return true;
    }
    return false;
  }

  //============================================================================

  /**
   * Deactivates the state.
   */
  void deactivate(Metadata data, Parameter parameter) {
    if(data.isActive(this)) {
      // Mark events as invalid
      Vector<EventQueueEntry> timeoutEvents = data.getData(this).timeoutEvents;
      if(timeoutEvents != null) {
        for (EventQueueEntry event : timeoutEvents) {
          event.invalid = true;
          /*
           * @FIXME
           * @See Java-Bug ID: 6268068. Proposed fix is to use:
           * removeAll(Collections.singletonList(event)).  
           *
           * Problem: Leads to ConcurrentModificationException. Final fix is in JDK6
           */
          statechart.timeoutEventQueue.remove(event);
        }
      }
      data.deactivate(this);
      if(exitAction != null) {
        exitAction.execute(data, parameter);
      }      
    }
  }

  //============================================================================

  /**
   * Dispatches the given event.
   */
  boolean dispatch(Metadata data, Event event, Parameter parameter) {
    for(int i = 0; i < transitions.size(); i++)  {
      if(transitions.get(i).execute(event, data, parameter)) {
        return true;
      }
    }
    return false;
  }
}
